src/box/field_def.c

```diff
@@ -164,7 +324,16 @@ field_type1_contains_type2(enum field_type type1, enum field_type type2)
 		mp_type1 = field_mp_type[type1];
 		mp_type2 = field_mp_type[type2];
 	}
-	return (mp_type1 & mp_type2) == mp_type2;
+
+	bool type1_contains_type2 = (mp_type1 & mp_type2) == mp_type2;
+	if (!type1_contains_type2)
+		return false;
+
+	int64_t min1 = field_type_min_value[type1];
+	int64_t min2 = field_type_min_value[type2];
+	uint64_t max1 = field_type_max_value[type1];
+	uint64_t max2 = field_type_max_value[type2];
+	return min1 <= min2 && max1 >= max2;
```

Comment:
IMO it's better not to set the limits for non-fixed types and instead check the `is_fixed` flag here:

```C
if (field_type_is_fixed_signed[type2]) {
	return !field_type_is_fixed_signed[type1] ||
		(field_type_min_value[type1] <= field_type_min_value[type2] &&
		 field_type_max_value[type1] >= field_type_max_value[type2]);
}
if (field_type_is_fixed_unsigned[type2]) {
	return (!field_type_is_fixed_signed[type1] &&
		!field_type_is_fixed_unsigned[type1]) ||
		field_type_max_value[type1] >= field_type_max_value[type2];
}
```

Reply:
This condition doesn't work for:
```
    not ok 41 - int8 can store values of unsigned
    not ok 42 - uint8 can store values of unsigned
    not ok 43 - int16 can store values of unsigned
    not ok 44 - uint16 can store values of unsigned
    not ok 45 - int32 can store values of unsigned
    not ok 46 - uint32 can store values of unsigned
    not ok 47 - int64 can store values of unsigned
    not ok 141 - int8 can store values of integer
    not ok 143 - int16 can store values of integer
    not ok 145 - int32 can store values of integer
    not ok 147 - int64 can store values of integer
```
I've fixed it.